assets/model_monitoring/components/src/model_data_collector_preprocessor/span_tree_utils.py

```diff
@@ -0,0 +1,168 @@
+
+    def __iter__(self) -> Iterator["SpanTreeNode"]:
+        """Iterate over current span and child spans."""
+        for child_span in self._children:
+            for span in child_span:
+                # print(f"iter func: child_span: {child_span}, span: {span}.")
+                yield span
+        yield self
+
+    def __lt__(self, other: "SpanTreeNode") -> bool:
+        """Compare by end_time in bisect.insort() for python3.8."""
+        return self.span_row.end_time < other.span_row.end_time
+
+
+class SpanTree:
+    """Spantree class."""
+
+    def __init__(self, spans: List[SpanTreeNode]) -> None:
+        """Spantree constructor to build up tree from span list."""
+        self.root_span = self._construct_span_tree(spans)
+        self._load_json_tree_lazy = False
+
+    @classmethod
+    def create_tree_from_json_string(cls, json_string: str, load_tree_lazy: bool = False) -> "SpanTree":
+        """Create SpanTree object from "root_span" json string."""
+        obj = cls.__new__(cls)
+        super(SpanTree, obj).__init__()
+        obj._load_json_tree_lazy = load_tree_lazy
+        # Default behavior is to load the whole tree from top level json string.
+        # TODO: Stretch goal will be to load the tree lazily one level at a time to save computation/space.
+        if not load_tree_lazy:
+            obj.root_span = obj._from_json_str_repr(json_string)
+        return obj
+
+    def show(self) -> None:
+        """Print to stdout a formatted representation of the Span Tree."""
+        if self.root_span is None:
+            return
+        self.root_span.show()
+
+    def to_json_str(self) -> str:
+        """Get tree structure as json string."""
+        if self.root_span is None:
+            return None  # type: ignore
+        return self._to_json_str_repr(self.root_span)
+
+    def _construct_span_tree(self, spans: List[SpanTreeNode]) -> SpanTreeNode:
+        """Build the span tree in ascending time order from list of all spans."""
+        # construct a dict with span_id as key and span as value
+        span_map = {span.span_id: span for span in spans}
+        for span in span_map.values():
+            parent_id = span.parent_id
+            if parent_id is None:
+                root_span = span
+            else:
+                parent_span = span_map.get(parent_id)
+                if parent_span is not None:
+                    parent_span.insert_child(span)
+        return root_span
+
+    def __iter__(self) -> Iterator[SpanTreeNode]:
+        """Iterate over the span tree in order."""
+        if self.root_span is None:
+            return
+        for span in self.root_span.__iter__():
+            yield span
+
+    def _from_json_str_repr(self, json_string: str) -> SpanTreeNode:
+        """Convert json string to SpanTree Node fully."""
+        output_node = SpanTreeNode.create_node_from_json_str(json_string)
+        child_subtree_nodes = []
+        for child in output_node.children:
```

Comment:
I believe we should move this children logic to SpanTreeNode

Reply:
In v2.